NIFI-11204: Add configurable retry logic for table commits in PutIceberg processor

mark-bathori opened the pull request:
<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

[NIFI-11204](https://issues.apache.org/jira/browse/NIFI-11204)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [ ] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ ] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [ ] Pull Request based on current revision of the `main` branch
- [ ] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 11
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

krisztina-zsihovszki reviewed (commented):
Thank you for the improvements, @mark-bathori!

I also noticed during the review that penalize is not called before FlowFile is sent to Failure relationship.
You may also consider sending provenance event for successful FlowFiles.

mark-bathori commented:
Thanks @krisztina-zsihovszki for the review. I addressed all your review comments and added the missing penalize calls before the failure relation transfers. I think the missing provenance event sending would be better to be added in a separate pull request. What do you think?

krisztina-zsihovszki commented on nifi-nar-bundles/nifi-iceberg-bundle/nifi-iceberg-processors/src/main/java/org/apache/nifi/processors/iceberg/PutIceberg.java:
```diff
@@ -113,6 +116,42 @@ public class PutIceberg extends AbstractIcebergProcessor {
 
+    static final PropertyDescriptor NUMBER_OF_COMMIT_RETRIES = new PropertyDescriptor.Builder()
+            .name("number-of-commit-retries")
+            .displayName("Number Of Commit Retries")
```
Done

krisztina-zsihovszki commented on nifi-nar-bundles/nifi-iceberg-bundle/nifi-iceberg-processors/src/main/resources/docs/org.apache.nifi.processors.iceberg.PutIceberg/additionalDetails.html:
```diff
@@ -0,0 +1,57 @@
+        <h3>Description</h3>
+        <p>
+            Iceberg is a high-performance format for huge analytic tables.
+            The PutIceberg processor is capable to push data into iceberg tables using different type of Iceberg catalog implementations.
```
Done

krisztina-zsihovszki commented on nifi-nar-bundles/nifi-iceberg-bundle/nifi-iceberg-processors/src/main/resources/docs/org.apache.nifi.processors.iceberg.PutIceberg/additionalDetails.html:
```diff
@@ -0,0 +1,57 @@
+                Number Of Commit Retries (default: 3) - Number of retries that the processor is going to try to commit the new data files.
+            </li>
+            <li>
+                Minimum Commit Wait Time (default: 100 ms) - Minimum time that the processor is going to wait before each commit attempts.
```
Done

krisztina-zsihovszki commented on nifi-nar-bundles/nifi-iceberg-bundle/nifi-iceberg-processors/src/main/resources/docs/org.apache.nifi.processors.iceberg.PutIceberg/additionalDetails.html:
```diff
@@ -0,0 +1,57 @@
+                Minimum Commit Wait Time (default: 100 ms) - Minimum time that the processor is going to wait before each commit attempts.
+            </li>
+            <li>
+                Maximum Commit Wait Time (default: 100 ms) - Maximum time that the processor is going to wait before each commit attempts.
```
Done

krisztina-zsihovszki commented:
Thank you fixing my review items, @mark-bathori .
It is OK for me that the provenance events will be added in a separate PR.

mark-bathori commented on nifi-nar-bundles/nifi-iceberg-bundle/nifi-iceberg-processors/src/main/java/org/apache/nifi/processors/iceberg/PutIceberg.java:
```diff
@@ -113,6 +116,42 @@ public class PutIceberg extends AbstractIcebergProcessor {
+            .required(true)
+            .defaultValue("30 min")
+            .addValidator(StandardValidators.TIME_PERIOD_VALIDATOR)
+            .build();
```
Thanks @turcsanyip for the review. I've modified the defaults to your recommended values.

turcsanyip reviewed (approved):
@mark-bathori Thanks for the changes!
@krisztina-zsihovszki Thank you for your review!

+1 LGTM
Merging to main and 1.x
